### PR TITLE
Add an Atomic Interface

### PR DESCRIPTION
**Pull request by bkueng**
We are moving towards more fine-grained locking and synchronization, and therefore we need need to ensure certain data accesses and modifications are atomic. This PR introduces a simple API based on GCC's atomic builtins (also available on clang).

I deliberately did not add type conversion and assignment operators, to make the operations explicit.

### CODE REVIEW

**Comment by LorenzMeier** [comment]
Very cool! I didn't do a line-by-line and will leave that to @dagar.

File: src/platforms/px4_module.h [hunk 1]
```diff
@@ -114,7 +115,7 @@ template<class T>
 class ModuleBase
 {
 public:
-	ModuleBase() = default;
+	ModuleBase() : _task_should_exit{false} {}
```
Comment (dagar) [inline src/platforms/px4_module.h]: No longer default?

**Comment by dagar** [comment]
Looks good so far. I was actually contemplating just pulling in LLVM libc++ or perhaps a subset as needed. https://github.com/PX4/libcxx/blob/nuttx-wip/include/atomic

File: src/platforms/px4_module.h (hunk 1 above)
Comment (bkueng) [reply src/platforms/px4_module.h]: This is due to a compiler bug: https://github.com/PX4/Firmware/pull/11328/commits/adca15520ae835477d80e1c4da2afbe7e76edf1c.
Ideally we should update these toolchains.

**Comment by dagar** [comment]
Why not px4::atomic<T> instead of px4_atomic_t<T>?

**Comment by dagar** [comment]
I'm hoping we won't have to consider anything other than sequentially consistent ordering. :grimacing:
Let's at least get in the habit of doing basic performance sanity checks when introducing this anywhere hot.

**Comment by bkueng** [comment]
> Why not px4::atomic instead of px4_atomic_t?

I had it as a struct first with a C API. I changed it accordingly.

> I'm hoping we won't have to consider anything other than sequentially consistent ordering.

So do I.

**Pull request merged**